### ISSUE TITLE
Order changes before payment without warning -> empty orders

### ISSUE THREAD

**Issue by TimvanDijkhuizen** [craftcms/commerce#1503]
**Description**
Out of stock items are getting removed from the order without any notification and the payment isn't cancelled. Because of this bug I'm sometimes seeing empty orders when single item orders are updated without warning. I've disabled empty orders, so this shouldn't be possible.

`PaymentsController::actionPay` retrieves the cart using `$order = $plugin->getCarts()->getCart(true);` which forces the cart to resave meaning it gets recalculated. The reason the message isn't showing is because the order is getting resaved before the old values are saved at line 143.

**To Reproduce**
How, and what happened?
1. Add a single item to your cart
2. Go to the payment step of the checkout
3. Set the stock of the item to 0 in the CP
4. Try to pay for the item
5. You will be allowed to pay and you end up with an empty order

**Expected behavior**
You're supposed to see a message telling you the order has been updated and payment should be cancelled.

**Additional info**
- Craft version: 3.3.20.1
- PHP version: 7.2.18
- Database driver & version: MySQL 5.7.26
- Plugins & versions: Commerce 2.2.14

**Additional context**
I'm using an older version bug looking at the code I think this bug is still present in Commerce 3.

**Comment by TimvanDijkhuizen** [comment]
Update: I've found a workaround for now.

I have added a hidden orderNumber field to my payment form to bypass the getCart call. I'm now seeing an error but my template wasn't showing the updated order. Passing true to the getCart call in my template (aka resaving it) fixed that. I'm not sure why I wasn't seeing the updated order before, I'm guessing cache (inside a service or Order) is the cause.

**Comment by lukeholder** [comment]
Fixed in 2b0f7ff05fb88b6b0a00a37a74443267de554825

**Comment by lukeholder** [comment]
Thanks for reporting that! I’ve just fixed it for the next release.

To get the fix early, change your `craftcms/commerce` requirement in composer.json to:

```json
"require": {
  "craftcms/commerce": "dev-develop#2b0f7ff05fb88b6b0a00a37a74443267de554825 as 3.1.8",
  "...": "..."
}
```

Then run `composer update`.
